torchaudio/models/wav2vec2/model.py

```diff
@@ -116,23 +116,162 @@ def forward(
+
+        final_proj (torch.nn.Module):
+
+        aux (torch.nn.Module or None, optional):
+            Auxiliary module. If provided, the output from encoder is passed to this module.
+    """  # noqa: E501
+    def __init__(
+        self,
+        feature_extractor: Module,
+        encoder: Module,
+        mask_generator: Module,
+        mask_embedding: Parameter,
+        label_embeddings: Parameter,
+        final_proj: Module,
+        aux: Optional[Module] = None,
+    ):
+        super().__init__(feature_extractor, encoder, aux)
+        self.feature_extractor = feature_extractor
+        self.encoder = encoder
+        self.mask_generator = mask_generator
+        self.mask_embedding = mask_embedding
+        self.label_embeddings = label_embeddings
+        self.final_proj = final_proj
+        self.aux = aux
+
+    def _compute_pred(
+        self,
+        proj_x: Tensor,
+        target: Tensor,
+        label_embeddings: Tensor,
+    ) -> Tensor:
+        logit_temp = 0.1
+        pos = torch.index_select(label_embeddings, 0, target.long())
+        negs = label_embeddings.unsqueeze(1).expand(-1, proj_x.size(0), -1)
+        neg_is_pos = (pos == negs).all(-1)
+        pos = pos.unsqueeze(0)
+        targets = torch.cat([pos, negs], dim=0)
+
+        logits = torch.cosine_similarity(proj_x.float(), targets.float(), dim=-1).type_as(proj_x)
+        logits /= logit_temp
+        if neg_is_pos.any():
+            logits[1:][neg_is_pos] = float("-inf")
+        logits = logits.transpose(0, 1)  # (num_x, num_cls+1)
+        return logits
+
+    def _get_padding_mask(self, x: Tensor, lengths: Tensor) -> Tensor:
+        batch_size, max_len, _ = x.shape
+        # create mask for padded elements and zero-out them
+        mask = torch.arange(max_len, device=lengths.device).expand(batch_size, max_len) >= lengths[:, None]
+        return mask
+
+    def forward(
+            self,
+            waveforms: Tensor,
+            labels: Tensor,
+            audio_lengths: Optional[Tensor] = None,
+    ) -> Tuple[Tensor, Optional[Tensor]]:
+        """Compute the sequence of probability distribution over labels.
+
+        Args:
+            waveforms (Tensor): Audio tensor of shape `(batch, frames)`.
+            labels (Tensor): Label for Pre-training or finetuning task.
+            audio_lengths (Tensor or None, optional):
+                Indicates the valid length of each audio in the batch.
+                Shape: `(batch, )`.
+                When the ``waveforms`` contains audios with different durations,
+                by providing ``lengths`` argument, the model will compute
+                the corresponding valid output lengths and apply proper mask in
+                transformer attention layer.
+                If ``None``, it is assumed that all the audio in ``waveforms``
+                have valid length. Default: ``None``.
+            label_lengths (Tensor or None, optional):
```

Comment:
This argument seems to be missing.

Reply:
Thanks. I will fix it. It seems `label_lengths` is not necessary, using `audio_lengths` helps creates the padding mask for labels.